code/drasil-code/lib/Language/Drasil/Code/Imperative/FunctionCalls.hs

```diff
@@ -35,7 +36,7 @@ getAllInputCalls = do
 
 -- | Generates a call to the function for reading inputs from a file.
 getInputCall :: (OOProg r) => GenState (Maybe (MSStatement r))
-getInputCall = getInOutCall "get_input" getInputFormatIns getInputFormatOuts
+getInputCall = getInOutCall (showChsStr GetInput) getInputFormatIns getInputFormatOuts
```

Comment:
Why is the first argument to `getInOutCall` still a `String` instead of being something renderable?

Reply:
Could you please explain what is meant by 'something renderable'?  Does the new `RenderChoicesStr` not now render the `get_input` string?

Reply:
Exactly: it renders it *here*. What if you waited to render things 'later'. Then point is that `GetInput` is more informative, because it is 'typed' and known.  So what you need is for this argument to be of class `RenderChoicesStr` rather than being rendered right away.

Reply:
`getInOutCall` is expecting a `Name` (`String`) as its first argument. It does not seem right to change the signature of `getInOutCall`, which is used in several places, and a `Name` seems like a good argument type. 

AM I wrong and should I change the signature of `getInOutCall` along with any `Name`'s that are passed into it, or should I leave the signature of `getInOutCall` as is and update Choices.hs to have `GetInput` be a `Name` (`String`)? 

I'm trying to understand if the problem is the code that I had submitted for this PR, or if the issue is deeper than that and this  requires an update of the current `getInOutCall` function and any code that is passed in as its arguments.

Reply:
That's exactly the discussion: should `getInOutCall` be some kind of very simple function that can't do any reasoning at all, or should it have the capability to do complex things because it knows more about its inputs?  I'm thinking the latter.

So yes, that would involve changing the signature of `getInOutCall` so that its first input be an instance of `RenderChoicesStr` instead of being a `Name`. I think of something of type `RenderChoicesStr a => a` as something that can have a name (in the future), but right now has a different, hopefully more informative, representation.

Reply:
I understand the high level idea what is wanted but am unsure of the implementation based on the above. Currently we have:

```
getInOutCall :: (OOProg r) => Name -> GenState [CodeVarChunk] -> 
  GenState [CodeVarChunk] -> GenState (Maybe (MSStatement r))
```

If I import RenderChoicesStr and update the function signature to the following, I can indeed  just use GetInput as the first input to the function.

```
getInOutCall :: (OOProg r, RenderChoicesStr a) => a -> GenState [CodeVarChunk] -> 
  GenState [CodeVarChunk] -> GenState (Maybe (MSStatement r))
```

But this creates more problems since a `Name` is needed for `fAppInOut` (in GenerateGOOL.hs).
`Couldn't match expected type ‘Name’ with actual type ‘a’`

Reply:
To follow up on this and the other comment above, do we want to do-away with `Name` in (all?) other functions as well? Otherwise, if you were suggesting in the other comment that we have _either_ `RenderChoicesStr` or `Name/String` then that would probably make things easier for now, but it is not clear to me yet at this point how to implement that.

Reply:
Just to record our conversation and apply it to here: we want to do away with `Name` as an argument to `getInOutCall` as a first step, along with whatever else is need to make that work. We don't want to do away with all of them in one big swoop. But we do want to (eventually) examine them all.

Also: for now, rather than going with the class-based, very general approach, let's go with just an algebraic datatype replacing `Name`.